### PR TITLE
REF: dispatch Block.fillna to putmask/where

### PR DESCRIPTION
**Pull request by jbrockmendel**
- [ ] closes #xxxx (Replace xxxx with the Github issue number)
- [ ] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [ ] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [ ] Added an entry in the latest `doc/source/whatsnew/vX.X.X.rst` file if fixing a bug or adding a new feature.

Among other things, we avoid re-calculating 'mask' when splitting.

### CODE REVIEW

File: pandas/core/internals/blocks.py [hunk 1]
```diff
@@ -1485,39 +1486,29 @@ def fillna(
...
-                        stacklevel=find_stack_level(),
-                    )
-                    raise
-                blk = self.coerce_to_target_dtype(value)
-                return blk.fillna(value, limit, inplace, downcast)
-
-            else:
+        if self.dtype.kind == "m":
+            try:
+                res_values = self.values.fillna(value, limit=limit)
```
Comment (jorisvandenbossche) [inline pandas/core/internals/blocks.py:1491]: What's the reason for putting this behind a `if self.dtype.kind == "m":` check?

Because this now will call `super().fillna()` for all EAs, but the parent `fillna()` implementation doesn't actually call `EA.fillna()`

(this was catched in the geopandas test suite)

Comment (jorisvandenbossche) [reply pandas/core/internals/blocks.py:1491]: Which future warning?

Let me rephrase my question: what's the reason for no longer calling EA.fillna() for all EAs?

Comment (jbrockmendel) [reply pandas/core/internals/blocks.py:1491]: > Which future warning?

The FutureWarning on L1494-1502 just below this.

> Let me rephrase my question: what's the reason for no longer calling EA.fillna() here for all EAs?

Primarily code-sharing, with the side-benefit that we may actually respect the "inplace" keyword (see TODO comment L1508)

Comment (jorisvandenbossche) [reply pandas/core/internals/blocks.py:1491]: Well, the point of the `EA.fillna()` method is that this gets called from the DataFrame/Series fillna, so that an EA can provide a specialized implementation. So I think we should keep that logic.

Comment (jbrockmendel) [reply pandas/core/internals/blocks.py:1491]: > so that an EA can provide a specialized implementation

does anyone (e.g. pyarrow) actually have a specialized implementation where it makes a real difference?  e.g. is more performant than going through putmask/where?

> the point of the EA.fillna() method is that this gets called from the DataFrame/Series fillna

I see "the point" of the EA interface in general as being to provide a minimal(ish) set of primitives from which we can  implement the Series/DataFrame(/Index!) API.  If Series.foobar can be implemented efficiently without EA.foobar, that is a reason to remove EA.foobar, not to use it unnecessarily.

(Though in this case we still use EA.fillna for pad/bfill/interpolate cases, so I'm not actually advocating getting rid of it)

Last, while the corrected 'inplace' behavior was not the motivator behind this change, I do think it is an important improvement.
